crypto/stack/stack.c

```diff
@@ -12,6 +12,34 @@
 #include "internal/numbers.h"
 #include <openssl/stack.h>
 #include <openssl/objects.h>
+#include <errno.h>
+
+/*
+ * The initial number of nodes in the array.
+ */
+static const int min_nodes = 4;
+
+/*
+ * Specify the array growth factor as a rational by defining a numerator
+ * and a denominator.  According to Andrew Koenig in his paper "Why Are
+ * Vectors Efficient?" from JOOP 11(5) 1998, this factor should be less
+ * than the golden ratio (1.618...).
+ *
+ * We use 3/2 = 1.5 for simplicty of calculation and overflow checking.
+ * Another option 8/5 = 1.6 allows for faster growth, although safe
+ * computation is more difficult.
+ *
+ * Two points are important:
+ *  1. The denominator must be smaller than the min_nodes value.
+ *  2. The fraction must be greater than unity.
+ *
+ * The first is because the division is performed first and truncating to
+ * zero is bad.  The second is so that the array does grow.  The truncation
+ * from the division happens before the multiplication, hence ratios close
+ * to unity could be problematic for some sizes (e.g. 6/5 with a size of 18).
+ */
+static const int growth_numerator = 3;
+static const int growth_denominator = 2;
```

Comment:
Question here is all [supported] compilers manage to optimize division as shift...

Reply:
I suppose this also answers the remark about `x+x/2`. I mean as long as one wants to keep the options open, one has to accept that `(x/2)*3` is less exact. Well, it's matter of taste, since assertion is that *any* factor less than 1.618 would do...

Reply:
Maybe have a define or function to get the next size, rather than two contants.

Reply:
I was aware that the division by two became a shift, the expansion shouldn't be a common occurrence so the cost of a division would be acceptable -- some compilers will change this to a multiplication (_Division by invariant integers using multiplication_ by Granlund and Montgomery).

I also realised that `(x/2)*3` is low by _one_ for odd `x`.  Correcting this will add complexity to the overflow checking and that's something I was trying to avoid -- it is difficult enough to reason about these checks and an error in them has consequences.  The latter is the main reason I avoided the 8/5 option which is arguably slightly better (being closer to the magic value Φ).

I'm happy to pull that bit of code out into a separate function.  It's only in `OPENSSL_sk_reserve` at the moment.  I wanted to avoid cluttering that function with a long comment about the rationale behind the ratio choice and the concerns this raises.

Reply:
Is the exact rounding a serious reason for worry here?

Reply:
> Is the exact rounding a serious reason for worry here?

No, but I'd say it's appropriate to explicitly discuss it, so that if "but-what-about" question is risen later, there is record of conscious discussion.

Reply:
The most accurate downward rounding has been done.  The off by one should be gone.